### PR TITLE
add import of tslint issues report

### LINKED ISSUE #610: Add import of tslint issues report
Thanks to https://jira.sonarsource.com/browse/MMF-1212 we can add support for import of issues from third-party tools. A new property holding tslint issue report should be added.
Note that feature is available only with SQ 7.2 (should be released in mid June) and on [SonarCloud](https://about.sonarcloud.io/)

### PR DESCRIPTION
**Pull request by vilchik-elena**
Fixes #610

### CODE REVIEW

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/TypeScriptPlugin.java [hunk 1]
```diff
@@ -73,6 +78,15 @@ public void define(Context context) {
         .category(TYPESCRIPT_CATEGORY)
         .multiValues(true)
         .build(),
+      PropertyDefinition.builder(TSLINT_REPORT_PATHS)
+        .defaultValue(TSLINT_REPORT_PATHS_DEFAULT_VALUE)
+        .name("TSLint Result Files")
```
Comment (saberduck) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/TypeScriptPlugin.java]: why "result" and not "report"?

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/externalreport/TslintReportSensor.java [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+  }
+
+  @Override
+  public void execute(SensorContext context) {
+    boolean externalIssuesSupported = context.getSonarQubeVersion().isGreaterThanOrEqual(Version.create(7, 2));
+    String[] tslintReportPaths = context.config().getStringArray(TSLINT_REPORT_PATHS);
+
+    if (tslintReportPaths.length == 0) {
+      return;
+    } else if (!externalIssuesSupported) {
```
Comment (saberduck) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/externalreport/TslintReportSensor.java]: we don't need `else if` because of `return`

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/externalreport/TslintReportSensor.java [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+      LOG.error("Current version of SonarQube doesn't support import of external issues (at least 7.2 required).");
```
Comment (saberduck) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/externalreport/TslintReportSensor.java]: "Current version" is bit ambiguous here. I would rephrase as "Import of external issues requires SonarQube 7.2 or greater."

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/externalreport/TslintReportSensor.java [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+  private void importTslintReport(File report, SensorContext context) {
+    LOG.info("Importing {}", report.getAbsoluteFile());
+
+    try (InputStreamReader inputStreamReader = new InputStreamReader(new FileInputStream(report), StandardCharsets.UTF_8)) {
+      TslintError[] tslintErrors = gson.fromJson(inputStreamReader, TslintError[].class);
+      for (TslintError tslintError : tslintErrors) {
+        saveTslintError(context, tslintError);
+      }
+    } catch (IOException e) {
+      LOG.error("No TSLint issues information will be saved because file cannot be found.", e);
```
Comment (saberduck) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/externalreport/TslintReportSensor.java]: I would remove "because file cannot be found", there might be other types of exception, better not to be too specifc

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/externalreport/TslintReportSensor.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    }
+  }
+
+  private void saveTslintError(SensorContext context, TslintError tslintError) {
+    if (activatedTslintKeys.contains(tslintError.ruleName)) {
+      return;
```
Comment (saberduck) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/externalreport/TslintReportSensor.java]: Maybe we can have `DEBUG` level log here, so user can investigate why issue is not imported, something like "Issue on line 42 skipped because the same issue will be reported by SonarTS"

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/externalreport/TslintReportSensor.java [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+    NewIssueLocation primaryLocation = newExternalIssue.newLocation()
+      .message(tslintError.failure)
+      .on(inputFile)
+      .at(getLocation(tslintError, inputFile));
+
+    newExternalIssue
+      .at(primaryLocation)
+      .forRule(RuleKey.of("tslint", tslintError.ruleName))
+      .type(BUG_TSLINT_RULES.contains(tslintError.ruleName) ? RuleType.BUG : RuleType.CODE_SMELL)
```
Comment (saberduck) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/externalreport/TslintReportSensor.java]: I thought that type will be based on `ruleSeverity` information from the report

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/TypeScriptPlugin.java (hunk 1 above)
Comment (vilchik-elena) [reply sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/TypeScriptPlugin.java]: @saberduck you are right, I will change it. TSLint docs uses "result", so I stuck to this word unconsciously :)

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/externalreport/TslintReportSensor.java (hunk 4 above)
Comment (vilchik-elena) [reply sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/externalreport/TslintReportSensor.java]: @saberduck indeed! i replaced it with `No TSLint issues information will be saved as report file can't be read.`

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/externalreport/TslintReportSensor.java (hunk 5 above)
Comment (vilchik-elena) [reply sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/externalreport/TslintReportSensor.java]: @saberduck great idea!

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/externalreport/TslintReportSensor.java [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+  private void saveTslintError(SensorContext context, TslintError tslintError) {
+    String tslintKey = tslintError.ruleName;
+
+    if (activatedRules.containsKey(tslintKey)) {
+      String message = "TSLint issue for rule '{}' is skipped because this rule is activated in your SonarQube profile for TypeScript (rule key in SQ {})";
+      LOG.debug(message, tslintKey, activatedRules.get(tslintKey));
+      return;
+    }
+
+    InputFile inputFile = context.fileSystem().inputFile(context.fileSystem().predicates().hasRelativePath(tslintError.name));
```
Comment (saberduck) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/externalreport/TslintReportSensor.java]: it would be nice to try to support absolute paths too
